EIP712/Implementation/EIP712.cs

```diff
@@ -0,0 +1,203 @@
+﻿using EIP712.Attributes;
+using EIP712.Exceptions;
+using EIP712.Utilities;
+using Nethereum.ABI.Encoders;
+using Nethereum.Signer;
+using Nethereum.Util;
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Numerics;
+using System.Reflection;
+using System.Text;
+
+namespace EIP712
+{
+    public static class EIP712
+    {
+
+        private static readonly byte[] _eip191Header = new byte[] { 0x19, 0x1 };
+        private static readonly Sha3Keccack _keccak = Sha3Keccack.Current;
+
+        #region PublicApi
+        /// <summary>
+        /// Encodes structured data according to EIP-712 specification
+        /// </summary>
+        /// <typeparam name="T">Structured data datatype</typeparam>
+        /// <param name="structure">Structured data to encode</param>
+        /// <param name="domain">EIP-712 domain</param>
+        /// <returns>Encoded data</returns>
+        public static byte[] Encode<T>(T structure, EIP712Domain domain) where T : class
+        {
+            if (domain == null)
+                throw new ArgumentNullException(nameof(structure));
+            if (domain == null)
+                throw new ArgumentNullException(nameof(domain));
+
+            byte[] domainSeparator = HashStruct(domain);
+            byte[] hash = HashStruct(structure);
+            byte[] result = ByteUtil.Merge(_eip191Header, domainSeparator, hash);
+            return result;
+        }
+
+        /// <summary>
+        /// Hash structured data 
+        /// </summary>
+        /// <typeparam name="T">Structured data datatype</typeparam>
+        /// <param name="structure">Structured data to hash</param>
+        /// <param name="domain">EIP-712 domain</param>
+        /// <returns>Keccak256 of encoded data</returns>
+        public static byte[] Hash<T>(T structure, EIP712Domain domain) where T : class
+        {
+            if (structure == null)
+                throw new ArgumentNullException(nameof(structure));
```

Comment:
Just for the future, document these exceptions in the `<exception>` XML comment section, so that consumers of the library can at least suspect the exception origin 😄